### ISSUE TITLE
Using IsCsrfTokenValid attribute with invalid token redirects user to login page

### ISSUE THREAD

**Issue by cyve** [symfony/symfony#57343]
### Symfony version(s) affected

7.1.1

### Description

Hi,
I tried the new `IsCsrfTokenValid` attribute on a route to handle a form. When the token is invalid, the user is redirected to the login page instead of an error page. Probably because `IsCsrfTokenValidAttributeListener` throws a `InvalidCsrfTokenException` witch extends `AuthenticationException`.

### How to reproduce

Add `IsCsrfTokenValid` attribute on a route handling a form
```php
#[Route('/add-to-cart', name: 'add_to_cart', methods: ['POST'])]
#[IsCsrfTokenValid('add_to_cart')]
public function __invoke(Request $request): Response
{
}
```

Load the page and wait for the token to expire (or generate an invalid token)
```html
<form method="post" action="{{ path('add_to_cart') }}">
  <input type="hidden" name="product" value="{{ product.id }}">
  <input type="hidden" name="_token" value="this_token_is_invalid">
</form>
```

### Possible Solution

Maybe we could throw a `BadRequestHttpException` instead of an `InvalidCsrfTokenException` in `IsCsrfTokenValidAttributeListener` ? But I guess there is a good reason for `InvalidCsrfTokenException` to extend `AuthenticationException`, so I can't really see the implications.
If this solution looks good to you, I can create a PR.

**Comment by eltharin** [comment]
I think InvalidCsrfTokenException must extends AccessDeniedException instaed of AuthenticationException to get a 403 response instead a login page redirection.

For the moment it's possible in a custom authenticator to set the start function as mentioned in https://github.com/symfony/symfony/issues/16026#issuecomment-1439565626

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by eltharin** [comment]
sure, 
https://github.com/eltharin/reproducer_symfony_57343.git
juste have to make a composer install and serve it, 
If you go to https://127.0.0.1:8000/test you will be redirect to /login (I don't agree with that but it's another discussion)
if you log with user@fr.fr / password and go to /test you are still redirected to /login
if you change InvalidCsrfTokenException extends  AccessDeniedException instaed of AuthenticationException you have an error page with Invalid CSRF Token 

do you want I make the PR ? (Bug or New Feature ? )

This symptome is joining issues : 
https://github.com/symfony/symfony/issues/16026
https://github.com/symfony/symfony/issues/54318
I'm working on a PR for add a security option to block the login redirect and show error if we are already logged for those who want to prevent these behavior.

**Comment by eltharin** [comment]
With reflexion, BadRequestHttpException  will be a better choice for extends, beacause a Csrf Error is not automaticly linked to an authentification

**Comment by eltharin** [comment]
What about the HTTP code ? 403 is for acces denied, Is really CSRF error an acces denied ? For a contact form can we speak about Access ?

**Comment by netonevess** [comment]
for a temporary solution, I did the following code:
[https://github.com/netonevess/Symfony72-csrf-response-modifyer/blob/main/InvalidTokenListener.php](https://github.com/netonevess/Symfony72-csrf-response-modifyer/blob/main/InvalidTokenListener.php)

My software was redirecting to the login page once an invalid CSRF using the new attribute #[IsCsrfTokenValid] and now a custom JSON message.

**Comment by eltharin** [comment]
Yeah I created a Exception Transformer near you made, But it can be better to be in core

**Comment by netonevess** [comment]
> Yeah I created a Exception Transformer near you made, But it can be better to be in core

please help us to improve that one. You can create a PR to this repo

**Comment by eltharin** [comment]
I already did the PR, juste waiting for someone valid it : https://github.com/symfony/symfony/pull/57622

**Comment by nicolas-grekas** [comment]
Maybe @yguedidi would like to have a look at this issue and the related PR proposal? (because #52961)

**Comment by eltharin** [comment]
friendly ping @yguedidi

**Comment by janderik** [comment]
This is a meaningful security concern. When CSRF token validation fails, the current behavior of redirecting to login can introduce subtle security issues.

**Why Redirecting to Login on CSRF Failure Is Problematic**

1. **Masking attack attempts:** A redirect to `/login` makes CSRF attacks indistinguishable from legitimate expired sessions in logs and monitoring. If an attacker is systematically probing for CSRF vulnerabilities, the 302 response hides the signal. A distinct `403 Forbidden` or `400 Bad Request` makes attack detection straightforward:

```
# Attacker probing CSRF — both look the same with redirect
GET /transfer  → 302 → /login  (expired session)
POST /transfer → 302 → /login  (CSRF failure)  ← indistinguishable

# With proper status code
POST /transfer → 403 CSRF token validation failed  ← clear signal
```

2. **Leaking authentication state:** The redirect to `/login` implicitly reveals that the user **is not authenticated** (or their session expired). For a CSRF attack where the attacker controls the referrer, this leak could be observable. More importantly, it means a CSRF failure is treated identically to an authentication failure — these are semantically different events.

3. **Breaking user experience for legitimate flows:** If a user double-submits a form or has a stale tab open, they should see a clear "resubmit with new token" message, not be dumped to a login screen where they may re-authenticate and lose context.

**Recommended Behavior (per OWASP CSRF Prevention Cheat Sheet)**

Per the [OWASP CSRF Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/CSRF_Prevention_Cheat_Sheet.html):

> "The application should return an HTTP 403 Forbidden status code when CSRF validation fails."

In Symfony, the appropriate response would be:

```php
// In the firewall / CSRF listener
public function onCsrfValidationFailure(): Response
{
    return new JsonResponse(
        ['error' => 'CSRF token validation failed'],
        Response::HTTP_FORBIDDEN  // 403
    );
}

// Or for form-based flows, a clear error page:
throw new AccessDeniedException('Invalid CSRF token. Please go back and try again.');
```

**Specific Symfony consideration:** Symfony's `CsrfTokenManager` validates tokens early in the request lifecycle. If the failure is in a form handler or controller, a `400 Bad Request` is also acceptable:

```php
// In a controller action
if (!$this->isCsrfTokenValid('transfer', $request->request->get('_token'))) {
    throw new BadRequestHttpException('Invalid CSRF token');
}
```

The key principle: **a CSRF failure is a security violation, not an authentication failure**. The response should clearly communicate this distinction to both the user and any monitoring infrastructure.

Would a PR changing the default to 403 be welcome? We'd need to ensure backward compatibility by making the behavior configurable.

**Comment by eltharin** [comment]
> The key principle: **a CSRF failure is a security violation, not an authentication failure**. The response should clearly communicate this distinction to both the user and any monitoring infrastructure.
> 
> Would a PR changing the default to 403 be welcome? We'd need to ensure backward compatibility by making the behavior configurable.

this PR exist since 2 years... https://github.com/symfony/symfony/pull/57622
